cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataUpgrader.java

```diff
@@ -0,0 +1,180 @@
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * Runs when Metadata Service starts to execute Metatdata Spec Update and Data Migration
+ */
+public class MetadataUpgrader {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MetadataUpgrader.class);
+  private static final int RETRY_DELAY = 2; // seconds
+  private static final RetryStrategy retryStrategy = RetryStrategies.fixDelay(RETRY_DELAY, TimeUnit.SECONDS);
+
+  private final ExistingEntitySystemMetadataWriter existingEntitySystemMetadataWriter;
+  private final MetadataStore metadataStore;
+  private final DatasetFramework datasetFramework;
+  private final NamespaceQueryAdmin namespaceQueryAdmin;
+
+  // A single thread executor to perform all upgrade related work
+  private ExecutorService executor;
+
+  @Inject
+  MetadataUpgrader(DatasetInstanceManager datasetInstanceManager,
+                   ExistingEntitySystemMetadataWriter existingEntitySystemMetadataWriter, MetadataStore metadataStore,
+                   DatasetFramework datasetFramework, NamespaceQueryAdmin namespaceQueryAdmin) {
+    this.existingEntitySystemMetadataWriter = existingEntitySystemMetadataWriter;
+    this.metadataStore = metadataStore;
+    this.datasetFramework = datasetFramework;
+    this.namespaceQueryAdmin = namespaceQueryAdmin;
+  }
+
+  public void executeUpgrade() {
+    // Upgrade the Specs inline
+    upgradeMetadataDatasetSpecs();
+
+    // Create a single thread executor using daemon threads which runs data-migration process.
+    executor = Executors.newSingleThreadExecutor(Threads.createDaemonThreadFactory("metadata-upgrader"));
+    executor.submit(new Callable<Void>() {
+      @Override
+      public Void call() throws Exception {
+        runUpgrade();
+        return null;
+      }
+    });
+  }
+
+  private void runUpgrade() throws Exception {
+    Predicate<Throwable> alwaysRetry = new Predicate<Throwable>() {
+      @Override
+      public boolean apply(@Nullable Throwable input) {
+        return true;
+      }
+    };
+    boolean upgradeRequired;
+    try {
+      upgradeRequired = Retries.callWithRetries(new Retries.Callable<Boolean, Exception>() {
+        @Override
+        public Boolean call() throws Exception {
+          return (metadataStore.isUpgradeRequired(MetadataScope.USER)
+            || metadataStore.isUpgradeRequired(MetadataScope.SYSTEM));
+        }
+      }, retryStrategy);
+    } catch (Exception e) {
+      LOG.error("Failed to check Metadata Dataset Upgrade status.", e);
+      throw e;
+    }
+
+    if (!upgradeRequired) {
+      LOG.info("Metadata Dataset Upgrade is not required.");
+      return;
+    }
+
+    LOG.info("Metadata DataSet Upgrade is reqiured.");
+    try {
+      Retries.callWithRetries(new Retries.Callable<Void, Exception>() {
+        @Override
+        public Void call() throws Exception {
+          // Run data migration
+          reIndex();
```

Comment:
should you retry the entire reIndex()? I think it would be better to retry individual batches. The reason is that this upgrade now runs in a live system. That means, CDAP is running and it may update metadata at any time. If it updates a row that coincides with the batch that you are currently reindexing, then your transaction will fail because that is a conflict. But if you retry just that batch, it will probably succeed. It would be pointless to abort the entire re-index and start over.